network/discovery/src/file.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use diem_config::{config::PeerSet, network_id::NetworkContext};
+use diem_logger::prelude::*;
+use diem_time_service::{Sleep, TimeService, TimeServiceTrait};
+use futures::{Future, Stream};
+use network::logging::NetworkSchema;
+use std::{
+    path::{Path, PathBuf},
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+pub struct FileStream {
+    network_context: Arc<NetworkContext>,
+    file_path: PathBuf,
+    time_service: TimeService,
+    interval_duration: Duration,
+    delay: Option<Pin<Box<Sleep>>>,
+}
+
+impl FileStream {
+    pub(crate) fn new(
+        network_context: Arc<NetworkContext>,
+        file_path: &Path,
+        interval_duration: Duration,
+        time_service: TimeService,
+    ) -> Self {
+        FileStream {
+            network_context,
+            file_path: file_path.to_path_buf(),
+            time_service,
+            interval_duration,
+            delay: None,
+        }
+    }
+}
+
+impl Stream for FileStream {
+    type Item = PeerSet;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Wait for delay, or add the delay for next call
+        if let Some(ref mut delay) = self.delay {
+            futures::ready!(Pin::new(delay).poll(cx));
+        }
+        self.delay = Some(Box::pin(self.time_service.sleep(self.interval_duration)));
+
+        match load_file(self.file_path.as_path()) {
+            Ok(peers) => Poll::Ready(Some(peers)),
+            Err(error) => {
+                error!(
+                    NetworkSchema::new(&self.network_context),
+                    "{} Failed to load file: {:?}", self.network_context, error
+                );
+                Poll::Ready(None)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum DiscoveryError {
+    IO(std::io::Error),
+    Parsing(String),
+}
+
+/// Loads a YAML configuration file
+fn load_file(path: &Path) -> Result<PeerSet, DiscoveryError> {
+    let contents = std::fs::read_to_string(path).map_err(DiscoveryError::IO)?;
```

Comment:
ideally this should use `tokio::fs::read_to_string`, which will run on the blocking threadpool so we don't block the async event loop.

you will have to change `FileStream` a bit though. To avoid all the tricky poll state machine stuff (and since this stream isn't polled very often) we can just do this:

```rust
struct FileStream {
    // ..
    updates: Pin<Box<Stream<Item = PeerSet>>>,
}

impl FileStream {
    fn new(
        // ..
        read_frequency: Duration,
        time_service: TimeService,
    ) -> Self {
        let updates = time_service
            .interval(read_frequency)
            .then(|_| load_file(path));
            .boxed();
        // ..
        Self {
            // ..
            updates,
        }
    }
}

impl Stream for FileStream {
    type Item = Result<PeerSet>;

    fn poll_next(/* .. */) -> Poll<Self::Item> {
        self.updates.poll_next(cx)
    }
}
```

Reply:
is the plan to do this in a follow up?

Reply:
The plan is to do this in a followup.  I spent some time trying to implement the above, and ran into a lot of rust issues.  I'd like to progress on my main task of onchain discovery, and delay this secondary stepping stone of a feature's improvements to a later time.